urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/geom/twodim/ops/SimplePolygonOp.java

```diff
@@ -0,0 +1,483 @@
+		BoundingBox bbox = iter.next().bbox();
+		while(iter.hasNext()){
+			bbox = bbox.union(iter.next().bbox());
+		}
+		BoundingBox[] bboxes = bbox.split();
+		
+		//split polygons
+		iter = polys.iterator();
+		Set<IntersectSimplePolygon>[] subs = new HashSet[4];//NOTE: for IntersectPolygon it use Object's equals() function. No problem
+		for(int i=0; i<4; i++){
+			subs[i] = new HashSet<IntersectSimplePolygon>();
+		}
+		while(iter.hasNext()){
+			IntersectSimplePolygon poly = iter.next();
+			BoundingBox b = poly.bbox();
+			for(int i=0; i<4; i++){
+				if(!(bboxes[i].intersect(b)instanceof Empty))
+					subs[i].add(poly);
+			}
+		}
+		
+		int maxSize = (int)Math.ceil(polys.size()*maxFactor);//NOTE: this factor must be less than 1. Think of identical polygons
+		if(subs[0].size()>maxSize || subs[1].size()>maxSize  || subs[2].size()>maxSize || subs[3].size()>maxSize){
+			//don't split it
+			result.add(polys);
+			return result;
+		}			
+		for(int i=0; i<4; i++){
+			if(subs[i].size()<minSize){//NOTE: don't call to save time
+				result.add(subs[i]);
+			}else{
+				result.addAll(split(subs[i]));
+			}
+		}
+		return result;
+	}
+	
+	
+	/*
+	 * Compute the union of p1 and p2.
+	 * Return a new polygon with same vertices if p1 == p2
+	 * Return a new polygon with vertices same with p1 if p1 contains p2;
+	 * Return a new polygon with vertices same with p2 if p2 contains p1;
+	 * Return the leftmost polygon if p1 and p2 disjoint.
+	 */
+	public static SimplePolygon union(SimplePolygon p1, SimplePolygon p2){
+		if(!(p1.type() instanceof ScaleType) || p1.type()!=p2.type() )
+			throw new GeomException("Input polygons must be exact and same type");
+		ScaleType type = (ScaleType)p1.type();
+		//find intersection points
+		IntersectSimplePolygon pi1 = new IntersectSimplePolygon(p1);
+		IntersectSimplePolygon pi2 = new IntersectSimplePolygon(p2);
+		pi1.intersect(pi2);
+		
+		//find left most point
+		IntersectPoint left = pi1.left();
+		if(pi2.left().compareTo(left)<0)
+			left = pi2.left();
+		
+		//take a tour and find all points		
+		List<IntersectPoint> points = new LinkedList<IntersectPoint>(); 
+		IntersectPoint point = left;
+		Point pre = Point.create((ScaleNumber)point.x().sub(1), (ScaleNumber)point.x());
+		Segment currSeg = new Segment(pre,point);//scale type
+		IntersectSegment nextSeg ;
+		do{
+			if(points.contains(point)){
+				throw new NonSimplePolygonException("Infinite loop, the input or the output is not simple polygon\n");
+			}
+			points.add(point);
+			nextSeg = point.advance(currSeg,true);
+			point = nextSeg.advance(point);
+			currSeg = nextSeg;
+		}while(!point.equals(left));		
+		return new SimplePolygon(toScale(points,type));
+	}
+	
+
+	/*
+	 * Compute the intersection of p1 and p2.
+	 * Return null if intersection is empty, point or segment
+	 * The result may not be correct if the intersection is non-simple polygon.
+	 */
+	public static SimplePolygon intersect(SimplePolygon p1, SimplePolygon p2){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 271-282 into a separate method.